File: algorithm/sorting/quickSort.js
function quickSort(arr, l=0, r=arr.length-1) {

	if (l >= r) return
	let i = l, j = r
	let pivot = arr[l]
	while (i != j) {
		//順序重要 先檢查 r 再檢查 i, r 往前跳與 l 重疊，不符合 i < j, 否則 i 會多算一格造成錯誤
		while (arr[j] > pivot && i < j) j--
		while (arr[i] <= pivot && i < j) i++
		if (i < j) {
			let tmp = arr[i]
			arr[i] = arr[j]
			arr[j] = tmp
		}
	}
	arr[l] = arr[i]
	arr[i] = pivot
	quickSort(arr, l, i - 1)
	quickSort(arr, i + 1, r)
}

let box = Array.from({length: 20}, () => Math.floor(Math.random() * 40))
console.log('original array:', box)
quickSort(box)
console.log('sorted   array:', box)

File: grad/sorting/quick_sort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#define N 6

void print_arr(int arr[]) {
	for (int i = 0; i < N; i++)
		printf("%d ", arr[i]);
	printf("\n");
}
void swap(int arr[], int a, int b) {
	int tmp = arr[a];
	arr[a] = arr[b];
	arr[b] = tmp;
}
void quick_sort(int arr[], int left, int right) {
	
	if (left >= right) return;

	int pivot = left;
	int l = left;
	int r = right;

	while (l != r) {
		while (arr[r] > arr[pivot] && l < r) r--;
		while (arr[l] <= arr[pivot] && l < r) l++;
		if (l < r) swap(arr, l, r);
	}
	
	swap(arr, pivot, l);
	pivot = l;
	quick_sort(arr, left, pivot - 1);
	quick_sort(arr, pivot + 1, right);
}
int main(int argc, char const *argv[]) {
	srand(time(NULL));
	int arr[N];
	for (int i = 0; i < N; i++)
		arr[i] = (rand()%50)+1;
	print_arr(arr);
	quick_sort(arr, 0, N-1);
	print_arr(arr);
	return 0;
}